src/main/java/io/prestosql/hadoop/ForwardingSocket.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.hadoop;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.SocketChannel;
+
+import static java.util.Objects.requireNonNull;
+
+class ForwardingSocket
+        extends Socket
+{
+    protected final Socket socket;
```

Comment:
nit: in the spirit of consistency with Guava `Forwarding*` classes, you'd have `protected abstract Socket delegate()` instead of a field

Reply:
I'm aware of that pattern, but it doesn't seem to provide any value here and makes the code longer. I went with the simpler `FilterInputStream` pattern.